File: JavaScript/7js 3/separateTypes.js
/*
|--------------------------------------------------------------------------
| Separate basic types
|--------------------------------------------------------------------------
|
*/

function separateTypes(input) {
   let obj = {};
   const num = input.filter(el => typeof el == "number");
   const str = input.filter(el => typeof el == "string");
   const bool = input.filter(el => typeof el == "boolean");
   if (num.length >= 1) obj.number = num;
   if (str.length >= 1) obj.string = str;
   if (bool.length >= 1) obj.boolean = bool;
   return obj;
}


separateTypes = a => a.reduce((x, y) => ({
   ...x,
   [q = typeof y]: [...x[q] || [], y]
}), {})


function separateTypes(input, data = ['number', 'string', 'boolean'], obj = {}) {
   let arr = data.map(x => input.filter(d => typeof d == x));
   for (let i = 0; i < arr.length; i++) {
      if (arr[i].length > 0) obj[data[i]] = arr[i];
   }
   return obj;
}


const separateTypes = arr => arr.reduce((acc, curr) => ({
   ...acc,
   [typeof curr]: acc[typeof curr] ? [...acc[typeof curr], curr] : [curr]
}), {})


console.log(separateTypes(['a', 1, 2, false, 'b']), {
   number: [1, 2],
   string: ['a', 'b'],
   boolean: [false]
});
console.log(separateTypes(['a', 1, 2]), {
   number: [1, 2],
   string: ['a']
});

File: Python/7py 3/separate_types.py
"""
|--------------------------------------------------------------------------
| Separate basic types
|--------------------------------------------------------------------------
|
"""

def separate_types(seq): 
   res = {}
   xint = list(filter(lambda i: type(i) == int, seq))
   xstr = list(filter(lambda i: type(i) == str, seq))
   xbool = list(filter(lambda i: type(i) == bool, seq))
   if len(xint) > 0: res[int] = xint
   if len(xstr) > 0: res[str] = xstr
   if len(xbool) > 0: res[bool] = xbool
   return res


separate_types = lambda S: {t:[e for e in S if type(e)==t] for t in {type(e) for e in S}}


def separate_types(seq):
   res = {}    
   for i in seq:
      res.setdefault(type(i), []).append(i)
   return res


def separate_types(seq): 
   res = {}
   for element in seq:
      if type(element) not in res:
         res[type(element)] = [element]
      else:
         res[type(element)].append(element)     
   return res


print(separate_types(['a', 1, 2, False, 'b']))
print(separate_types(['a', 1, 2]))